google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1beta2/StreamWriter.java

```diff
@@ -922,41 +922,43 @@ public void onError(Throwable t) {
         }
         inflightBatch = this.inflightBatches.poll();
       }
-      try {
-        if (isRecoverableError(t)) {
-          try {
-            if (streamWriter.currentRetries < streamWriter.getRetrySettings().getMaxAttempts()
-                && !streamWriter.shutdown.get()) {
-              streamWriter.refreshAppend();
-              LOG.info("Resending requests on transient error:" + streamWriter.currentRetries);
-              streamWriter.writeBatch(inflightBatch);
-              synchronized (streamWriter.currentRetries) {
-                streamWriter.currentRetries++;
-              }
-            } else {
-              inflightBatch.onFailure(t);
-              abortInflightRequests(t);
-              synchronized (streamWriter.currentRetries) {
-                streamWriter.currentRetries = 0;
-              }
+      streamWriter.messagesWaiter.release(inflightBatch.getByteSize());
+      if (isRecoverableError(t)) {
+        try {
+          if (streamWriter.currentRetries < streamWriter.getRetrySettings().getMaxAttempts()
+              && !streamWriter.shutdown.get()) {
+            synchronized (streamWriter.currentRetries) {
+              streamWriter.currentRetries++;
             }
-          } catch (IOException | InterruptedException e) {
-            LOG.info("Got exception while retrying.");
-            inflightBatch.onFailure(e);
-            abortInflightRequests(e);
+            LOG.info(
+                "Try to reestablish connection due to transient error: "
+                    + t.toString()
+                    + " retry times: "
+                    + streamWriter.currentRetries);
+            streamWriter.refreshAppend();
+            LOG.info("Resending requests on after connection established");
+            streamWriter.writeBatch(inflightBatch);
+          } else {
+            inflightBatch.onFailure(t);
+            abortInflightRequests(t);
             synchronized (streamWriter.currentRetries) {
               streamWriter.currentRetries = 0;
             }
           }
-        } else {
-          inflightBatch.onFailure(t);
-          abortInflightRequests(t);
+        } catch (IOException | InterruptedException e) {
+          LOG.info("Got exception while retrying: " + e.toString());
+          inflightBatch.onFailure(e);
+          abortInflightRequests(e);
```

Comment:
can we add test coverage for catching these lines?

Reply:
good point. I tried this morning but couldn't figure out a good way to hit the code. I will play around a little more, but could we proceed with this first. Datalfow is waiting for a fix.

Reply:
SG